python-package/xgboost/_typing.py

```diff
@@ -1,20 +1,27 @@
 """Shared typing definition."""
 import ctypes
 import os
-from typing import Optional, List, Any, TypeVar, Union
+from typing import Optional, Any, TypeVar, Union, Type, Sequence
 
 # os.PathLike/string/numpy.array/scipy.sparse/pd.DataFrame/dt.Frame/
 # cudf.DataFrame/cupy.array/dlpack
+import numpy as np
+
 DataType = Any
 
 # xgboost accepts some other possible types in practice due to historical reason, which is
 # lesser tested.  For now we encourage users to pass a simple list of string.
-FeatureNames = Optional[List[str]]
+FeatureInfo = Optional[Sequence[str]]
```

Comment:
Should we change to `Sequence[str]` here, and add `Optional` in the signatures when needed? Some functions require `feature_names` to be _not_ None, for example, https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/core.py#L2562-L2563

Reply:
I think mypy suggests using `Sequence` for function input and more specific type like `List` and `Tuple` as function return types. https://mypy.readthedocs.io/en/stable/generics.html?highlight=variance#variance-of-generic-types

Reply:
Not sure if we're talking about the same thing... I meant to use `FeatureInfo = Sequence[str]` here, and when referring to it, use `Optional[FeatureInfo]`
I've also drafted a commit to better illustrate my idea: https://github.com/bridgream/xgboost/commit/5e3d4aff364a959f784ff316aa66b52484b20448

This isn't something major, though...